File: backend/src/main/java/de/learnlib/alex/testsuites/services/TestService.java
/*
 * Copyright 2016 TU Dortmund
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.learnlib.alex.testsuites.services;

import de.learnlib.alex.auth.entities.User;
import de.learnlib.alex.config.entities.BrowserConfig;
import de.learnlib.alex.data.entities.ExecuteResult;
import de.learnlib.alex.data.entities.Symbol;
import de.learnlib.alex.learning.services.connectors.ConnectorContextHandler;
import de.learnlib.alex.learning.services.connectors.ConnectorContextHandlerFactory;
import de.learnlib.alex.learning.services.connectors.ConnectorManager;
import de.learnlib.alex.learning.services.connectors.VariableStoreConnector;
import de.learnlib.alex.testsuites.entities.Test;
import de.learnlib.alex.testsuites.entities.TestCase;
import de.learnlib.alex.testsuites.entities.TestCaseResult;
import de.learnlib.alex.testsuites.entities.TestSuite;
import de.learnlib.alex.testsuites.entities.TestSuiteResult;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * The service that executes tests.
 */
@Service
public class TestService {

    /** Factory to create a new ContextHandler. */
    private ConnectorContextHandlerFactory contextHandlerFactory;

    /**
     * Constructor.
     *
     * @param contextHandlerFactory The injected {@link ConnectorContextHandlerFactory}
     */
    @Inject
    public TestService(ConnectorContextHandlerFactory contextHandlerFactory) {
        this.contextHandlerFactory = contextHandlerFactory;
    }

    /**
     * Execute a test suite.
     *
     * @param user          The user that executes the test suite.
     * @param testSuite     The test suite that is being executed.
     * @param browserConfig The config for the web driver.
     * @return
     */
    public TestSuiteResult executeTestSuite(User user, TestSuite testSuite, BrowserConfig browserConfig) {
        TestSuiteResult tsResult = new TestSuiteResult(testSuite, 0L, 0L);

        for (Test test : testSuite.getTests()) {
            if (test instanceof TestCase) {
                TestCaseResult result = executeTestCase(user, (TestCase) test, browserConfig);
                tsResult.add(result);
                tsResult.getResults().put(test.getId(), result);
            } else {
                TestSuiteResult r = executeTestSuite(user, (TestSuite) test, browserConfig);
                tsResult.add(r);
                tsResult.getResults().put(test.getId(), r);
            }
        }

        return tsResult;
    }

    /**
     * Executes a test case.
     *
     * @param user          The user that executes the test suite.
     * @param testCase      The test case that is being executed.
     * @param browserConfig The config for the web driver.
     * @return
     */
    public TestCaseResult executeTestCase(User user, TestCase testCase, BrowserConfig browserConfig) {
        final ConnectorContextHandler ctxHandler = contextHandlerFactory.createContext(user, testCase.getProject(),
                                                                                       browserConfig);
        ctxHandler.setResetSymbol(new Symbol());

        final long startTime = System.currentTimeMillis();
        final ConnectorManager connectors = ctxHandler.createContext();

        // set variables from the test case
        final VariableStoreConnector variableStore = connectors.getConnector(VariableStoreConnector.class);
        testCase.getVariables().forEach(variableStore::set);

        // execute the test case
        final List<ExecuteResult> outputs = testCase.getSymbols().stream()
                .map(s -> s.execute(connectors))
                .collect(Collectors.toList());

        connectors.dispose();
        final long time = System.currentTimeMillis() - startTime;

        final List<String> sulOutputs = outputs.stream().map(ExecuteResult::getOutput).collect(Collectors.toList());

        final List<Symbol> failedSymbols = new ArrayList<>();
        final List<String> failureMessageParts = new ArrayList<>();

        for (int i = 0; i < outputs.size(); i++) {
            final ExecuteResult output = outputs.get(i);
            if (!output.isSuccessful()) {
                final Symbol symbol = testCase.getSymbols().get(i);
                failedSymbols.add(symbol);
                failureMessageParts.add(symbol.getName() + ":" + output.getOutput());
            }
        }

        final String failureMessage = failureMessageParts.isEmpty() ? null : String.join(", ", failureMessageParts);

        return new TestCaseResult(testCase, sulOutputs, failedSymbols.size() == 0, time, String.join(", ", failureMessage));
    }

}

File: frontend/src/main/javascript/src/js/services/test.service.js
/*
 * Copyright 2016 TU Dortmund
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * The service for test cases and test suites.
 */
export class TestService {

    /**
     * Constructor.
     *
     * @param $uibModal
     * @param {SymbolResource} SymbolResource
     * @param {TestResource} TestResource
     */
    // @ngInject
    constructor($uibModal, SymbolResource, TestResource) {
        this.$uibModal = $uibModal;
        this.SymbolResource = SymbolResource;
        this.TestResource = TestResource;
    }

    /**
     * Prepare tests for export.
     *
     * @param {object} tests The tests to export.
     * @return {*}
     */
    exportTests(tests) {
        const prepareTestCase = (testCase) => {
            deleteProperties(testCase);

            testCase.steps = testCase.steps.map((step) => {
                if (step.type === 'symbol') {
                    step.symbol = step.symbol.name;
                    step.parameterValues.forEach(value => {
                        delete value.id;
                        delete value.parameter.id;
                    });
                }
                return step;
            });
        };

        const prepareTestSuite = (testSuite) => {
            deleteProperties(testSuite);
            testSuite.tests.forEach(test => {
                if (test.type === 'case') {
                    prepareTestCase(test);
                } else {
                    prepareTestSuite(test);
                }
            });
        };

        const deleteProperties = (test) => {
            delete test.id;
            delete test.project;
            delete test.user;
            delete test.parent;
            delete test._selected;
        };

        for (const test of tests) {
            if (test.type === 'case') {
                prepareTestCase(test);
            } else {
                prepareTestSuite(test);
            }
        }

        return tests;
    }

    /**
     * Import tests.
     *
     * @param {number} projectId The id of the project.
     * @param {object} testsToImport The tests to import.
     * @param {number} parentId The id of the parent test suite.
     */
    importTests(projectId, testsToImport, parentId) {
        return this.SymbolResource.getAll(projectId)
            .then((symbols) => {
                const tests = JSON.parse(JSON.stringify(testsToImport));

                const mapTestCaseSymbols = (testCase) => {
                    testCase.steps = testCase.steps.map((step) => {
                        if (step.type === 'symbol') {
                            const sym = symbols.find(s => s.name === step.symbol);
                            if (sym) {
                                step.symbol = sym.id;
                            }
                        }
                        return step;
                    });
                };

                const prepareTestCase = (testCase, parent) => {
                    testCase.project = projectId;
                    testCase.parent = parent;
                    mapTestCaseSymbols(testCase);
                };

                const prepareTestSuite = (testSuite, parent) => {
                    testSuite.project = projectId;
                    testSuite.parent = parent;
                    testSuite.tests.forEach(test => {
                        if (test.type === 'case') {
                            prepareTestCase(test, null);
                        } else {
                            prepareTestSuite(test, null);
                        }
                    });
                };

                for (let test of tests) {
                    if (test.type === 'case') {
                        prepareTestCase(test, parentId);
                    } else {
                        prepareTestSuite(test, parentId);
                    }
                }

                return this.TestResource.createMany(projectId, tests);
            });
    }

    /**
     * Opens the test import modal.
     *
     * @param {object} testSuite The test suite where the imported tests should be imported.
     */
    openImportModal(testSuite) {
        return this.$uibModal.open({
            component: 'testsImportModal',
            resolve: {
                modalData: () => ({test: testSuite})
            }
        }).result;
    }
}
